### DISCUSSION TITLE
Native 2FA in Filament Panels (TOTP)

### DISCUSSION THREAD

**Discussion by Jacotheron** [filamentphp/filament#7981] [Ideas]
I would like to be able to enable 2FA on specific panels.

The way I see it, all websites web-services these days should provide a 2FA option for any user that needs to log in.

The way I think it should be implemented is as 2 methods added to the panel configuration (similar to how the registration method is implemented):
` ->twofactor(closure|bool)`
This one is basically to enable/disable the 2fa step. The closure can receive the current user, to for example, enable it only for certain users.

`->require_twofactor(closure|bool)`
This one will enable (if not already enabled) the 2fa, and if the closure returns true, require the user to setup 2fa immediately if it is not yet set up.

For most standard users, in the standard panels, providing 2fa as an option is good; however, typically on the Admin panels, we would like to require 2fa so that it is harder to get hacked.

Perhaps this implementation can also be made extendible (for example having a 3rd method, or 2nd attribute on the 1st method for the 2fa driver) - making it easier for 3rd parties to add a new driver (for example to use hardware keys, push notifications etc).

I currently wait for this functionality to upgrade to v3, as the 2fa package I have implemented, is not yet compatible. This made me think that having native/1st party support for this will be beneficial to a lot of people - and make it easier to build more secure websites. With such functionality, it remains the developer's choice to enable it (as a valid reason to not enable it is when it is only accessed locally).

**Comment by danharrin** [comment]
This plugin supports 2FA in v3 - https://github.com/jeffgreco13/filament-breezy

**Reply by Jacotheron** [reply]
I seem to have missed this part. Thank you.
